File: restChat.js
// Rest based chat client
// Jim Skon 2022
// Kenyon College

var baseUrl = 'http://44.202.89.194:5005';
var state="off";
var myname="";
var email="";
var password="";
var color="";
var username="";

var inthandle;
var masterUserList =[];
var activeUserList =[];

/* Start with text input and status hidden */
document.getElementById('chatinput').style.display = 'none';
document.getElementById('status').style.display = 'none';
document.getElementById('leave').style.display = 'none';
// Action if they push the join button
document.getElementById('login-btn').addEventListener("click", (e) => {
	join();
})
// Action if they push the register button
document.getElementById('register-btn').addEventListener("click", (e) => {
    register();
})
/* Set up buttons */
document.getElementById('leave-btn').addEventListener("click", leaveSession);
document.getElementById('send-btn').addEventListener("click", sendText);
document.getElementById('color-toggle').addEventListener("click", toggleMode);

// Watch for enter on message box
document.getElementById('message').addEventListener("keydown", (e)=> {
    if (e.code == "Enter") {
	e.preventDefault();
    sendText();
    clearTextbox();
    return false;
    }   
});

// function for toggling the css on the page between light and dark mode.
function toggleMode() {
	document.getElementById('chatBox').classList.toggle('dark-mode-chatbox');
}

// Call function on page exit
window.onbeforeunload = leaveSession;

// Registers a new user by sending their information to the server
function register() {
    username = document.getElementById('register-name').value;
    email = document.getElementById('register-email').value;
    password = document.getElementById('register-pass').value;
    var temp = document.getElementById("userColor").value;
	color = temp.slice(1);
    if (password.length < 6) {
	alert("Password too short, try again.");    
    }
    fetch(baseUrl+'/chat/register/'+username+'/'+email+'/'+password+'/'+color, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data =>completeRegister(data))
    .catch(error => {
        {alert("Error: Something went wrong:"+error);}
    })
}

// Handle the registration response from the server. Only registers if there exists no overlapping username/email.
function completeRegister(results) {
	var status = results['status'];
	if (status != "success") {
		alert("Username or email already exists!");
		leaveSession();
		return;
	}
	var user = results['user'];
	console.log("Register:"+user);
}

// Handle the response after attempting to join the chat
function completeJoin(results) {
	var status = results['status'];
	if (status != "success") {
		alert("User already exists");
		leaveSession();
		return;
	}
	var user = results['user'];
	console.log("Join:"+user);
	startSession(user);
}

// Joins the chat by sending the username and password to the server
function join() {
	myname = document.getElementById('yourname').value;
	mypassword = document.getElementById('yourpass').value;
	fetch(baseUrl+'/chat/join/'+myname+'/'+mypassword, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data =>completeJoin(data))
    .catch(error => {
        {alert("Error: Something went wrong:"+error);}
    })
}

function completeSend(results) {
	var status = results['status'];
	if (status == "success") {
		console.log("Send succeeded")
	} else {
		alert("Error sending message!");
	}
}

//function called on submit or enter on text input
function sendText() {
    var message = document.getElementById('message').value;
    console.log("Send: "+myname+":"+message);
	fetch(baseUrl+'/chat/send/'+myname+'/'+message, {
        method: 'get'
    })

    .then (response => response.json() )
    .then (data =>completeSend(data))
    .catch(error => {
        {alert("Error: Something went wrong:"+error);}
    })
    clearTextbox();    
}

// Clears up the textbox once the message is sent
function clearTextbox() {
    document.getElementById('message').value = "";
}

// Process the messages and user list
function completeFetch(result) {
	names = result["userlist"];
	names.forEach(function (m,i) {
		name = m["name"] + ", ";
		if  (masterUserList.includes(name) == false) {
		masterUserList.push(name);
		}
    });
	activeUserList=[];
	activeusers = result["activelist"];
	activeusers.forEach(function (m,i) {
		user = m["name"] + ", ";
		activeUserList.push(user);
		
	});
    messages = result["messages"];
	messages.forEach(function (m,i) {
		name = m['user'];
		message = m['message'];
		color = m['color'];
		document.getElementById('chatBox').innerHTML +=
	    	"<font color="+color+">" + name + ": </font>" + message + "<br />";
	});
	updateChatMembers();
}

/* Check for new messaged */
function fetchMessage() {
	fetch(baseUrl+'/chat/fetch/'+myname, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data =>completeFetch(data))
    .catch(error => {
        {console.log("Server appears down");}
    })  	
}

// Update the list of chat members
function updateChatMembers() {
	document.getElementById('members').innerHTML = ""
    masterUserList.forEach(function (element,i) {
        var user = element;
        if (activeUserList.includes(user)) {
            document.getElementById('members').innerHTML +=
            "<font color='green'>" + user + " " +"</font> ";
        }
        else {
            document.getElementById('members').innerHTML +=
            "<font color='#ff0000'>" + user + " " + "</font>";
        } 
    });
}
/* Functions to set up visibility of sections of the display */
function startSession(name){
    state="on";
    
    document.getElementById('yourname').value = "";
    document.getElementById('register').style.display = 'none';
    document.getElementById('user').innerHTML = "User: " + name;
    document.getElementById('chatinput').style.display = 'block';
    document.getElementById('members').style.display = 'block';
    document.getElementById('status').style.display = 'block';
    document.getElementById('leave').style.display = 'block';
    /* Check for messages every 500 ms */
    inthandle=setInterval(fetchMessage,500);
}

function completeLogout(result){
   // something to remove the user from the ActiveUsers list/array but not just client side
var status = result['status'];
	if (status == "success") {
		console.log("Logout successful")
	} else {
		alert("Error logging out!");
	}
}

// Logs out the current user from the chat
function logout(){
    fetch(baseUrl+'/chat/logout/'+myname, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data => completeLogout(data))
    .catch(error => {
        {console.log("Server appears down");}
    })
}

// Handle leaving the chat session and the features following, such as hiding elements and stopping sending messages
function leaveSession(){
	logout();
	state="off";
    document.getElementById('yourname').value = "";
    document.getElementById('register').style.display = 'block';
    document.getElementById('user').innerHTML = "";
    document.getElementById('chatinput').style.display = 'none';
    document.getElementById('members').style.display = 'none';
    document.getElementById('status').style.display = 'none';
    document.getElementById('leave').style.display = 'none';
	clearInterval(inthandle);
}

File: restChat.cpp
//
//  namesAPI.cc - a microservice demo program
//
// James Skon
// Kenyon College, 2022
//

#include <iostream>
#include <fstream>
#include <map>
#include <algorithm>
#include "httplib.h"
#include "user.h"

using namespace httplib;
using namespace std;

const int port = 5005;

// Takes a username, message, message map, color
// and constructs a JSON-formatted string containing the user's information and the message,
// and then this JSON string to the message list of all users in the message map.
void addMessage(string username, string message, map<string,vector<string>> &messageMap, string color) {
	/* iterate through users adding message to each */
	string jsonMessage = "{\"user\":\""+username+"\",\"message\":\""+message+"\",\"color\":\""+color+"\"}";
	for (auto userMessagePair : messageMap) {
		username = userMessagePair.first;
		messageMap[username].push_back(jsonMessage);
	}
}

// Takes a username, message map, masterUserList, activeUserList, and user map
// Returns a JSON-formatted string with the messages for the given user, masterUserList, and activeUserList
string getMessagesJSON(string username, map<string,vector<string>> &messageMap,vector<string> &masterUserList,vector<string> &activeUserList, map<string,user> &userMap) {
	/* retrieve json list of messages for this user along with a userlist*/
	bool first = true;
	string user;
	string result = "{\"messages\":[";
	for (string message :  messageMap[username]) {
		if (not first) result += ",";
		result += message;
		first = false;
	}
	result+="],";
  messageMap[username].clear();
  first = true;
  result += "\"userlist\":[";
  for (int i=0; i<masterUserList.size(); i++) {
    if (not first) result += ",";
	user = masterUserList[i];
    result += "{\"name\":""\"";
    result += user +"\"}";
    first = false;
  }
  result+= "],";
  string activeList = "\"activelist\":[";
  for (auto pair : userMap){
	string activeuser = pair.first;
	if (pair.second.getActive() && find(activeUserList.begin(),activeUserList.end(), activeuser) == activeUserList.end()) {
		activeUserList.push_back(pair.first);
	}
  }
  first = true;
  for (int i=0; i<activeUserList.size(); i++) {
	  if (not first) activeList += ",";
	  user = activeUserList[i];
	  activeList += "{\"name\":""\"";
	  activeList += user +"\"}";
	  first = false;
  }
	activeUserList.clear();
	result+=activeList;
	result += "]}";
	return result;
}

// Defines various endpoints for operations
int main(void) {
  Server svr;
  int nextUser=0;
  map<string,vector<string>> messageMap;
  vector<string> masterUserList;
  vector<string> activeUserList;
  map<string, user> userMap;
	
  /* "/" just returnsAPI name */
  svr.Get("/", [](const Request & /*req*/, Response &res) {
    res.set_header("Access-Control-Allow-Origin","*");
    res.set_content("Chat API", "text/plain");
  });


  // Registers a new user with the given username, email, password, and color
  svr.Get(R"(/chat/register/(.*)/(.*)/(.*)/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
    string username = req.matches[1];
    string email = req.matches[2];
    string password = req.matches[3];
    string color = req.matches[4];
    bool active = true;
    string result;
    cout << username << email << password << color << endl;
    bool emailUnique = true;
    for (auto pair : userMap){
	   if (email == pair.second.getEmail()){
		   emailUnique = false;
       cout << "email not unique" << endl;
	   } 
    }
    if (userMap.count(username) || !emailUnique) {
    	cout << "exists" << endl;
      result = "{\"status\":\"exists\"}";
    } else {
    	// Add user to messages map
        user newUser(username, email, password, color, active);
        cout << "user created" << endl;
        userMap[username]=newUser;
        masterUserList.push_back(username);
	result = "{\"status\":\"success\",\"user\":\""+username+"\",\"email\":\""+email+"\",\"password\":\""+password+"\",\"color\":\""+color+"\"}";
    }
	res.set_content(result, "text/json");
  });

  // Logs in a user with the given username and password
  svr.Get(R"(/chat/join/(.*)/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
    string username = req.matches[1];
    string password = req.matches[2];
    string result;
    vector<string> empty;
    cout << username << " joins" << endl;
    
    // Check if username and password matches with registered user
    if (userMap.count(username)) {
      if (password==userMap[username].getPassword()) {
        messageMap[username]=empty;
		result = "{\"status\":\"success\",\"user\":\"" + username + "\"}";
      }
    }
    else {
      result = "{\"status\":\"exists\"}";
    }
	userMap[username].setActive(true);
    res.set_content(result, "text/json");
  });

   // Sends a message from the given username's user
   svr.Get(R"(/chat/send/(.*)/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string username = req.matches[1];
	string message = req.matches[2];
	string result; 
	
    if (!messageMap.count(username)) {
    	result = "{\"status\":\"baduser\"}";
	} else {
		string userColor = userMap[username].getColor();
		addMessage(username,message,messageMap,userColor);
		result = "{\"status\":\"success\"}";
	}
    res.set_content(result, "text/json");
  });
  
   //Fetches messages and user lists for the given username's user.
   svr.Get(R"(/chat/fetch/(.*))", [&](const Request& req, Response& res) {
    string username = req.matches[1];
    res.set_header("Access-Control-Allow-Origin","*");
    string resultJSON = getMessagesJSON(username,messageMap,masterUserList,activeUserList,userMap);
    res.set_content(resultJSON, "text/json");  
    });

  // Returns a JSON-formatted string with a status of "success" and the given name
  svr.Get(R"(/chat/hello/(.*))", [&](const Request& req, Response& res) {
    string name = req.matches[1];
    res.set_header("Access-Control-Allow-Origin","*");
    string resultJSON = "{\"status\":\"success\",\"name\":\""+name+"\"}";
    res.set_content(resultJSON, "text/json");
  });

  // Logs out the user with the given name.
  svr.Get(R"(/chat/logout/(.*))", [&](const Request& req, Response& res) {
    string name = req.matches[1];
    res.set_header("Access-Control-Allow-Origin","*");
	userMap[name].setActive(false);
    string resultJSON = "{\"status\":\"success\",\"name\":\""+name+"\"}"; 
	res.set_content(resultJSON, "text/json");
  });

  cout << "Server listening on port " << port << endl;
  svr.listen("0.0.0.0", port);

}
